### PR TITLE
Backend and service changes to allow tagging a build

### PR DESCRIPTION
**Pull request by lilida**
The changes include:

1. The database changes required for the new tags table. The meta_info stores a Json string to let the table can be extended to other usage.

2. A new restful service api (/tags) including Create, Delete and Read.  (Update may come later)
	Create:
		It can be used for tagging a particular build.
	Delete:
		Delete by id
	Read:
		The retrieving interface accepts the same set of query parameters as what are used for retrieving builds info and share the same code to builds. Then it associates the build tag to these builds and return the result. The association algorithm is in BuildTagsManagerImpl. It firstly sorted tags by commit date. Then do a binary search to find where to start the search. Then it uses either the tag on the same commit or the most recent tag on the same repo and branch.

@sbaogang @yujunglo

### CODE REVIEW

File: deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/resource/Tags.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.validation.Valid;
+import javax.ws.rs.*;
+import javax.ws.rs.core.*;
+import java.net.URI;
+import java.util.List;
+
+@Path("/v1/tags")
```
Comment (sbaogang) [inline deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/resource/Tags.java]: can we add swagger annotate for both class and methods? you can find examples in other resource classes.

File: deploy-service/common/src/main/java/com/pinterest/deployservice/ServiceContext.java [hunk 2]
```diff
@@ -76,6 +76,9 @@
     private PlacementDAO placementDAO;
     private SpotAutoScalingDAO spotAutoScalingDAO;
 
+
```
Comment (sbaogang) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/ServiceContext.java]: remove empty lines

File: deploy-service/common/src/main/java/com/pinterest/deployservice/bean/TagBean.java [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+ * `target_type` varchar(30) NOT NULL,
+ * `target_id` varchar(64) NOT NULL,
+ * `target_name` varchar(64),
+ * `operator` varchar(64) NOT NULL,
+ * `created_date` bigint(20) NOT NULL
+ * `comments` varchar(256),
+ * `meta_info` text,
+ * `is_active` bit default 1,
+ * PRIMARY KEY tag_infosKEY (`id`),
+ * INDEX `target_id_idx` (`target_id`, `is_active`),
```
Comment (sbaogang) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/bean/TagBean.java]: can we prefix the index with table name, such as tags_****

File: deploy-service/common/src/main/java/com/pinterest/deployservice/bean/TagTargetType.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.pinterest.deployservice.bean;
+
+public enum TagTargetType {
+    Build,
+    Environment
```
Comment (sbaogang) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/bean/TagTargetType.java]: Can we add Teletraan or Global, so we could temporarily disable the deploy globally

File: deploy-service/common/src/main/java/com/pinterest/deployservice/bean/TagBean.java [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+ * `id` varchar(30) NOT NULL,
+ * `value` varchar(30) NOT NULL,
+ * `target_type` varchar(30) NOT NULL,
+ * `target_id` varchar(64) NOT NULL,
+ * `target_name` varchar(64),
+ * `operator` varchar(64) NOT NULL,
+ * `created_date` bigint(20) NOT NULL
+ * `comments` varchar(256),
+ * `meta_info` text,
+ * `is_active` bit default 1,
```
Comment (sbaogang) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/bean/TagBean.java]: I am not sure we want this, seems to me it complicates the case

File: deploy-service/common/src/main/java/com/pinterest/deployservice/bean/TagValue.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.pinterest.deployservice.bean;
+
+
+public enum TagValue {
```
Comment (sbaogang) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/bean/TagValue.java]: Should/could we allow arbitrary tags?

File: deploy-service/common/src/main/java/com/pinterest/deployservice/dao/BuildDAO.java [hunk 7]
```diff
@@ -1,12 +1,12 @@
 /**
  * Copyright 2016 Pinterest, Inc.
- *
+ * <p>
```
Comment (sbaogang) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/dao/BuildDAO.java]: remove for consistency

File: deploy-service/common/src/main/java/com/pinterest/deployservice/dao/BuildDAO.java [hunk 8]
```diff
@@ -49,9 +53,19 @@
...
 
+    List<BuildBean> getBuildsFromIds(Collection<String> ids) throws Exception;
+
     // Get total number of builds by build name
     long countBuildsByName(String buildName) throws Exception;
 
     // Get all unused builds whose publish time is before timeThreshold
     void deleteUnusedBuilds(String buildName, long timeThreshold, long numOfBuilds) throws Exception;
+
+    List<BuildBean> get(String scmCommit,
```
Comment (sbaogang) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/dao/BuildDAO.java]: the format looks very different

File: deploy-service/common/src/main/java/com/pinterest/deployservice/db/DBBuildDAOImpl.java [hunk 9]
```diff
@@ -1,12 +1,12 @@
 /**
  * Copyright 2016 Pinterest, Inc.
- *
+ * <p>
```
Comment (sbaogang) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/db/DBBuildDAOImpl.java]: remove

File: deploy-service/common/src/main/java/com/pinterest/deployservice/db/DBBuildDAOImpl.java [hunk 10]
```diff
@@ -25,50 +26,59 @@
...
+import java.util.Collection;
 import java.util.List;
 
 public class DBBuildDAOImpl implements BuildDAO {
+
+    private static final int DEFAULT_SIZE = 100;
+
     private static final String INSERT_BUILD_TEMPLATE =
-        "INSERT INTO builds SET %s";
+            "INSERT INTO builds SET %s";
```
Comment (sbaogang) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/db/DBBuildDAOImpl.java]: double check style settings

File: deploy-service/common/src/main/java/com/pinterest/deployservice/exceptions/InvalidInputException.java [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.pinterest.deployservice.exceptions;
+
+
+public class InvalidInputException extends Exception {
```
Comment (sbaogang) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/exceptions/InvalidInputException.java]: so far we have been use more generic exceptions, I will say unless we want to treat this exception differently, such as REST call return different status code etc. or the caller should handle this exception differently, I would suggest to use the generic ones for now.

File: deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/resource/Tags.java [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+        if (ret == null) {
+            throw new TeletaanInternalException(Response.Status.NOT_FOUND,
+                    String.format("TagInfo %s does not exist.", id));
+        }
+        return ret;
+    }
+
+    @GET
+    @Path("/builds")
+    public List<BuildTagBean>  getBuildsWithTags(@QueryParam("commit") String scmCommit,
```
Comment (sbaogang) [inline deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/resource/Tags.java]: I am not sure we want to expose BuildTagBean, we should discuss

File: deploy-service/common/src/main/java/com/pinterest/deployservice/bean/TagBean.java [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+
+import java.beans.Transient;
+
+/**
+ * CREATE TABLE `tags` (
+ * `id` varchar(30) NOT NULL,
+ * `value` varchar(30) NOT NULL,
+ * `target_type` varchar(30) NOT NULL,
+ * `target_id` varchar(64) NOT NULL,
+ * `target_name` varchar(64),
```
Comment (sbaogang) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/bean/TagBean.java]: do we need this?

File: deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/resource/Tags.java (hunk 1 above)
Comment (lilida) [inline deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/resource/Tags.java]: Added

File: deploy-service/common/src/main/java/com/pinterest/deployservice/bean/TagBean.java (hunk 3 above)
Comment (lilida) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/bean/TagBean.java]: Done

File: deploy-service/common/src/main/java/com/pinterest/deployservice/ServiceContext.java (hunk 2 above)
Comment (lilida) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/ServiceContext.java]: Done

File: deploy-service/common/src/main/java/com/pinterest/deployservice/bean/TagTargetType.java (hunk 4 above)
Comment (lilida) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/bean/TagTargetType.java]: Added TELETRAAN now

File: deploy-service/common/src/main/java/com/pinterest/deployservice/dao/BuildDAO.java (hunk 8 above)
Comment (lilida) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/dao/BuildDAO.java]: Fixed

File: deploy-service/common/src/main/java/com/pinterest/deployservice/db/DBBuildDAOImpl.java (hunk 9 above)
Comment (lilida) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/db/DBBuildDAOImpl.java]: Done

File: deploy-service/common/src/main/java/com/pinterest/deployservice/db/DBBuildDAOImpl.java (hunk 10 above)
Comment (lilida) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/db/DBBuildDAOImpl.java]: Applied the new style

File: deploy-service/common/src/main/java/com/pinterest/deployservice/exceptions/InvalidInputException.java (hunk 11 above)
Comment (lilida) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/exceptions/InvalidInputException.java]: Use generic or existing now

File: deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/resource/Tags.java (hunk 12 above)
Comment (lilida) [inline deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/resource/Tags.java]: Refactored based on our conversation

**Comment by sbaogang** [comment]
r+, thanks for the changes!

**Pull request merged**